proposals/text/0000-evaluation-haystack-2.md

```diff
@@ -0,0 +1,223 @@
+
+When evaluating a `Pipeline` we could also override its private `_run_component` function to evaluate every node it will run. This will 100% work for our implementation of `Pipeline`. If a user tries to evaluate a `Pipeline` that reimplements its own `run` method it might not be able to evaluate each `Component`. I believe this a worthy risky tradeoff.
+
+Overriding `_run_component` would also give us the chance to simulate optimal component outputs. `eval` could also accept an optional `simulated_output` dictionary containing the outputs of one or more `Component` that are in the `Pipeline`. It would look similar to this:
+
+```
+simulated_output = {
+  "component_name": {"answer": "120"},
+  "another_component_name": {"metadata": {"id": 1}}
+}
+```
+
+Another alternative would be to use the `expected_output` also to simulate outputs of intermediate components.
+
+#### Tracking progress
+
+To track progress we can also go on another direction. We could return partials results while iterating our inputs:
+
+```
+def eval(runnable, inputs, expected_outputs):
+    result = EvaluationResult(runnable, inputs, {}, expected_outputs)
+    for input_ in inputs:
+      output = runnable.run(input_)
+      result.append_output(output)
+      yield result
+```
+
+Or return some progress percentage and only in the end the final result.
+
+```
+def eval(runnable, inputs, expected_outputs):
+    outputs = []
+    total = len(inputs)
+    for i, input_ in enumerate(inputs):
+      output = runnable.run(input_)
+      outputs.append(output)
+      yield 100 * (i / total), None
+
+    yield 100, EvaluationResult(runnable, inputs, outputs, expected_outputs)
+```
+
+### `EvaluationResult`
+
+`EvaluationResult` won't contain partial metrics anymore as it does in v1.x but it will keep track of all the information used by `eval()`.
+This way we'll be able to save all the necessary information to a single file. That information will probably be:
+
+- Serialized Pipeline or Component
+- Inputs
+- Expected outputs
+- Actual outputs
+
+This data should be serializable to string so that it can be saved to file and loaded back whenever necessary.
+We shouldn't expect all input and output data to implement serialization methods like `to_dict` and `from_dict` like `Pipeline` and `component`s do, so we probably should find an alternative to handle serialization of types that don't. An unsafe option would be the use of `pickle`, but that's dangerous as it could lead to malicious code being executed.
+
+Given the above information we should be able to implement a single method to calculate predeterminated metrics or even custom ones.
+Known metrics could be defined as an enum to ease discoverability and documentation.
+
+```
+class Metric(Enum):
```

Comment:
Is it a deliberate API design choice that this is an Enum, or it's an implementation detail?

Reply:
Implementation detail, could be anything really.

I used an `Enum` just cause it made type hints easier to read .